src/main/java/io/cdap/plugin/http/source/common/http/HttpClient.java

```diff
@@ -59,29 +65,63 @@ public HttpClient(BaseHttpSourceConfig config) {
     } else {
       requestBody = null;
     }
+
+    if (config.getServiceAccountEnabled()) { // Service account authent used
+      try {
+        serviceAccountTokenHandler = new ServiceAccountTokenHandler(
+                config.getServiceAccountScope(),
+                config.getServiceAccountPrivateKeyJson()
+        );
+      } catch (Exception e) {
+        LOG.error("Error while getting Service Account Access token " + e);
+      }
+    }
+  }
+
+  public CloseableHttpResponse executeHTTP(String uriStr) throws IOException {
+    return executeHTTP(uriStr, 1);
   }
 
   /**
    * Executes HTTP request with parameters configured in plugin config and returns response.
    * Is called to load every page by pagination iterator.
    *
-   * @param uri URI of resource
+   * @param uriStr URI of resource
+   * @param retries Number of retries in case of error
    * @return a response object
    * @throws IOException in case of a problem or the connection was aborted
    */
-  public CloseableHttpResponse executeHTTP(String uri) throws IOException {
+  public CloseableHttpResponse executeHTTP(String uriStr, int retries) throws IOException {
     // lazy init. So we are able to initialize the class for different checks during validations etc.
     if (httpClient == null) {
       httpClient = createHttpClient();
+    } else {
+      if (config.getServiceAccountEnabled()) { // Service account authent used
+        if (serviceAccountTokenHandler.tokenExpireSoon()) {
+          httpClient = createHttpClient(); // Recreate client with fresh token
+        }
+      }
     }
 
-    HttpEntityEnclosingRequestBase request = new HttpRequest(URI.create(uri), config.getHttpMethod());
+    URI uri = URI.create(uriStr);
+
+    HttpEntityEnclosingRequestBase request = new HttpRequest(uri, config.getHttpMethod());
 
     if (requestBody != null) {
       request.setEntity(requestBody);
     }
 
-    return httpClient.execute(request);
+    CloseableHttpResponse response = httpClient.execute(request);
+
+    if (retries > 0 && config.getServiceAccountEnabled()) {
+      HttpResponse httpResponse = new HttpResponse(response);
+      if (httpResponse.getStatusCode() == 401) { // If 'Unhauthorized' error, recreate client with fresh token
```

Comment:
I don't think this retry logic is needed as we're recreating the client with a fresh token if it's going to expire soon (line 101)